orchagent/muxorch.cpp

```diff
@@ -480,70 +447,240 @@ bool MuxCable::isIpInSubnet(IpAddress ip)
+        if (tnh == SAI_NULL_OBJECT_ID)
+        {
+            SWSS_LOG_INFO("Null NH object id, retry for %s", peer_ip4_.to_string().c_str());
+            return false;
+        }
+
+        return nbr_handler_->disable(tnh);
     }
 }
 
-void MuxNbrHandler::update(IpAddress ip, string alias, bool add)
+void MuxCable::updateNeighbor(NextHopKey nh, bool add)
 {
+    sai_object_id_t tnh = mux_orch_->getNextHopTunnelId(MUX_TUNNEL, peer_ip4_);
+    nbr_handler_->update(nh, tnh, add, state_);
     if (add)
     {
-        neighbors_.add(ip);
-        if (!alias.empty() && alias != alias_)
+        mux_orch_->addNexthop(nh, mux_name_);
+    }
+    else
+    {
+        mux_orch_->removeNexthop(nh);
+    }
+}
+
+void MuxNbrHandler::update(NextHopKey nh, sai_object_id_t tunnelId, bool add, MuxState state)
+{
+
+    SWSS_LOG_INFO("Neigh %s on %s, add %d, state %d",
+                   nh.ip_address.to_string().c_str(), nh.alias.c_str(), add, state);
+
+    if (add)
+    {
+        if (!nh.alias.empty() && nh.alias != alias_)
+        {
+            alias_ = nh.alias;
+        }
+
+        if (neighbors_.find(nh.ip_address) != neighbors_.end())
         {
-            alias_ = alias;
+            return;
+        }
+
+        switch (state)
+        {
+        case MuxState::MUX_STATE_INIT:
+            neighbors_[nh.ip_address] = SAI_NULL_OBJECT_ID;
+            break;
+        case MuxState::MUX_STATE_ACTIVE:
+            neighbors_[nh.ip_address] = gNeighOrch->getLocalNextHopId(nh);
+            enable(false);
+            break;
+        case MuxState::MUX_STATE_STANDBY:
+            neighbors_[nh.ip_address] = tunnelId;
+            disable(tunnelId);
+            break;
+        default:
+            SWSS_LOG_NOTICE("State '%s' not handled for nbr %s update",
+                             muxStateValToString.at(state).c_str(), nh.ip_address.to_string().c_str());
+            break;
         }
     }
     else
     {
-        neighbors_.remove(ip);
+        MuxCableOrch* mux_cb_orch = gDirectory.get<MuxCableOrch*>();
+        IpPrefix pfx = nh.ip_address.to_string();
+
+        /* if current state is standby, remove the tunnel route */
+        if (state == MuxState::MUX_STATE_STANDBY)
+        {
+            remove_route(pfx);
+            mux_cb_orch->removeTunnelRoute(nh);
+        }
+        neighbors_.erase(nh.ip_address);
     }
 }
 
-bool MuxNbrHandler::enable()
+bool MuxNbrHandler::enable(bool update_rt)
 {
     NeighborEntry neigh;
+    MuxCableOrch* mux_cb_orch = gDirectory.get<MuxCableOrch*>();
 
-    auto it = neighbors_.getIpAddresses().begin();
-    while (it != neighbors_.getIpAddresses().end())
+    auto l_fn = [&] (NextHopKey nh_key, uint32_t val) {
+        SWSS_LOG_INFO("Inc ref for %s by %d", nh_key.ip_address.to_string().c_str(), val);
+        for (uint32_t cnt = 0; cnt < val; ++cnt)
+        {
+            gNeighOrch->increaseNextHopRefCount(nh_key);
```

Comment:
can we increase the refcnt by value instead of looping here?

Reply:
This is existing API and would need to define new if we have to increase by value. Also, considering the number of routes, prefer we can use this loop instead of defining new.

Reply:
that will be cleaner, correct?